test/UnitTestFramework/UnitTestFramework.js

```diff
@@ -167,8 +191,120 @@ var testRunner = function testRunner() {
+                        if (!passed) helpers.writeln(asyncTest.error[i]);
+                        helpers.writeln(passed ? "PASSED" : "FAILED");
+                    }
+                    !passed || passedTestCount++;
+                }
+                if (_verbose || executedTestCount != passedTestCount) {
+                    helpers.writeln("Summary of tests: total executed: ", executedTestCount,
+                      "; passed: ", passedTestCount, "; failed: ", executedTestCount - passedTestCount);
+                } else {
+                    helpers.writeln("pass");
+                }
+            });
+        },
+
+        runAsyncTest: function runAsyncTest(testIndex, testName, testBody) {
+            ///<summary>Runs test body catching all exceptions.<br/>
+            /// Result: prints LOADED/LOADING FAILED to the output.
+            ///</summary>
+            function logTestNameIf(b) {
+                if (b) {
+                    helpers.writeln("*** Loading test #", executedTestCount + 1, " (", testIndex, "): ", testName);
+                }
+            }
+
+            logTestNameIf(_verbose);
+
+            asyncTest.testIndex = testIndex;
+            asyncTest.promise[testIndex] = [];
+            asyncTest.resolve[testIndex] = [];
+
+            var isSuccess = true;
+            try {
+                testBody(testIndex);
+            } catch (ex) {
+                var message = ex.stack || ex.message || ex;
+                logTestNameIf(!_verbose);
+                helpers.writeln("Test threw exception: ", message);
+                isSuccess = false;
+            }
+            if (isSuccess) {
+                if (_verbose) {
+                    helpers.writeln("LOADED");
+                }
+                ++passedTestCount;
+            } else {
+                helpers.writeln("LOADING FAILED");
+                testRunner.asyncTestErr(testIndex, "LOADING FAILED");
+            }
+            ++executedTestCount;
+        },
+
+        asyncTestBegin : function asyncTestBegin(testIndex, testCount) {
+            _asyncMode = true;
+            asyncTest.testIndex = testIndex;
+        },
+
+        asyncTestEnd : function asyncTestEnd(testIndex, testCount) {
+            _asyncMode = false;
+            asyncTest.resolve[testIndex][testCount]();
+        },
+
+        async : function async(source, shouldFail) {
+            var testIndex = asyncTest.testIndex;
+            if (!shouldFail) {
+                var testCount = asyncTest.count[testIndex]++;
+                var promise = new Promise((resolve, reject) => {
+                    asyncTest.resolve[testIndex].push(resolve);
+                });
+                asyncTest.promise[testIndex].push(promise);
+                return 'testRunner.asyncTestBegin(' + testIndex + ", " + testCount + ');' + source + ";\ntestRunner.asyncTestEnd(" + testIndex + ", " + testCount + ");";

```

Comment:
Here is why we add stuff to the source, remind me if I am missing something:
WScript.LoadModule() loads and parses module code synchronously but executes it asynchronously. Ideally we will need it to return a promise and resolve/reject it later depending on the parsing and the execution. Essentially making it a full-fledged async API. This unit test framework change tries to bridge the gap by handling the promise part for APIs like this that are 1) asynchronous in nature, and 2) not tracking its execution through promise. At this point, only the loaded code that knows exactly when it ends can resolve the promise.


---
In reply to: [110764912](https://github.com/Microsoft/ChakraCore/pull/2822#discussion_r110764912) [](ancestors = 110764912)